add support for basic authentication

schrej opened the pull request:
**What's this PR do?**

This PR adds support for HTTP Basic Authentication.
Electron fires a `login` event whenever Basic Authentication is triggered within a webview. This event now gets handled by showing a authentication dialog.

**Any background context you want to provide?**

We have a special 2 Factor Authentication proxy in place, that uses Basic Authentication. If we want to use Zulip Desktop, we need it to support Basic Authentication as well.

**Screenshots?**

![image](https://user-images.githubusercontent.com/1710904/86244826-d2937380-bba8-11ea-8f6f-c8551efc4bf5.png)

**You have tested this PR on:**
  - [ ] Windows
  - [ ] Linux/Ubuntu
  - [x] macOS

I've noticed the CSS of the app is pretty messy. It would be a good refactoring task to streamline it, especially regarding form input styling.

schrej committed:
add support for basic authentication

akashnimare commented:
> We have a special 2 Factor Authentication proxy in place, that uses Basic Authentication. If we want to use Zulip Desktop, we need it to support Basic Authentication as well.

Is there any way I can see/check how it looks? Need to reproduce this auth case in order to better understand this PR. @andersk check this out please.

andersk commented on app/renderer/js/main.ts:
```diff
@@ -1040,6 +1042,24 @@ class ServerManagerView {
+		// We show an authentication dialog and return the username and password.
+		app.on('login', async (event, webContents, details, authInfo, callback) => {
+			// This doesn't seem to work, so it's done in app/main/index.ts as well.
+			event.preventDefault();
```
The reason it doesn’t work is that you’re accessing `app` via `remote`, which doesn’t preserve synchronicity. We want to get rid of `remote` (#915). So assuming we need to do this in the renderer process, we should instead have the main process `login` handler contact the renderer process using IPC, probably using the same `makeRendererCallback` mechanism that the permission request handler uses.

schrej replied:
I've found PRs regarding replacing webview with BrowserView. That might make it easier/possible to attach the event listener directly to that as well. I tried it with the webview but didn't get it to work.

I have no experience with electron, so I probably missed something.

andersk reviewed (commented):
Does this stop one Zulip organization from popping up a login dialog over a different Zulip organization? I’m worried this could be abused for phishing.

Also, this appears to just hide the used dialogs without ever removing them, which will leak memory and slow the app down over time.

schrej commented:
> Is there any way I can see/check how it looks? Need to reproduce this auth case in order to better understand this PR.

It's just [http basic authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#Basic_authentication_scheme). The easiest is probably a reverse proxy that requests http basic auth in front of a zulip instance.

Our 2FA setup simply expects a password and a TOTP concatenated in the password field.

> Does this stop one Zulip organization from popping up a login dialog over a different Zulip organization? I’m worried this could be abused for phishing.

Good point! It might be hard to limit it to a single tab. Alternatively we could show the host that's requesting authentication in the dialog.

> Also, this appears to just hide the used dialogs without ever removing them, which will leak memory and slow the app down over time.

It does only hide them, but future Dialogs use the same div to inject the dialog. It gets overwritten every time and doesn't leak memory. The idea was to allow showing other Dialogs as well in the future if necessary.

schrej committed:
improve basic authentication logic

replace electron remote with ipc calls
one dialog per tab
show origin in the authentication dialog

schrej commented:
Ok, I've implemented the suggestions now:

I'm using IPC calls instead of remote. As it's asynchronous I'm storing the callbacks in index.ts until a response occurs. They don't get cleaned up in case no response arrives (which shouldn't happen), but as there can only be one callback per WebContents that should be ok.

The dialog now shows the origin of the authentication request.
![image](https://user-images.githubusercontent.com/1710904/86359146-f02c1000-bc70-11ea-96bc-6b426a18a817.png)

There is one dialog per tab now. It's only visible when the tab is visible. The `Dialog` objects only hold a reference to `<body>` as long as they're not in use. They add a `<div>` to `<body>` to show a dialog and clean it up when it's dismissed.

schrej committed:
use makeRendererCallback instead of own logic

also add a helper function to extract the callback on renderer side

schrej commented:
Missed the remark regarding `makeRendererCallback`, now using that instead. I've also implemented a (cryptic looking) helper function that extracts the callback id and passes a callback function as well as the remaining arguments to the event handler.
